src/hip_hcc.cpp

```diff
@@ -139,6 +143,103 @@ std::atomic<int> g_lastShortTid(1);
 std::vector<ProfTrigger> g_dbStartTriggers;
 std::vector<ProfTrigger> g_dbStopTriggers;
 
+#ifdef WITH_HOSTCALL
+//=================================================================================================
+// Hostcall global state; destructed on application exit
+//=================================================================================================
+
+class hostcallState {
+    amd_hostcall_consumer_t *consumer;
+    std::unordered_multimap<const void*, void*> deviceQueues;
+    std::unordered_map<void*, void*> buffers;
+    std::mutex _mutex;
+
+    // Functions with suffix _internal do not try to obtain a
+    // lock. They should be called from a context where the critical
+    // data is already locked.
+    void removeBuffer_internal(void *queue) {
+        auto buffer = findBuffer_internal(queue);
+        if (!buffer)
+            return;
+        amd_hostcall_deregister_buffer(consumer, buffer);
+        buffers.erase(queue);
+
+        tprintf(DB_SYNC, "discarded hostcall buffer %p for queue %p\n", buffer, queue);
+    }
+
+    void* findBuffer_internal(void *queue) {
+        if (buffers.find(queue) == buffers.end())
+            return nullptr;
+        return buffers[queue];
+    }
+
+public:
+    hostcallState() : consumer(nullptr) {}
+
+    ~hostcallState() {
+        if (consumer) {
```

Comment:
should we lock before destroy the consumer?

Reply:
> 
> 
> should we lock before destroy the consumer?

Do you mean to say that some other HIP thread might be accessing the "consumer" variable when we destroy it? I guess it would be safer to do that, but not really necessary in this implementation. The global state is destroyed only when hip_hcc.so is unloaded, so it is safe to assume that no other thread is still accessing this global state.